packages/astro/src/server/get-auth.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  type AuthObject,
+  AuthStatus,
+  type SignedInAuthObject,
+  signedInAuthObject,
+  type SignedOutAuthObject,
+  signedOutAuthObject,
+} from '@clerk/backend/internal';
+import { decodeJwt } from '@clerk/backend/jwt';
+import type { APIContext } from 'astro';
+
+import { getSafeEnv } from './get-safe-env';
+import { getAuthKeyFromRequest } from './utils';
+
+type AuthObjectWithoutResources<T extends AuthObject> = Omit<T, 'user' | 'organization' | 'session'>;
+
+export type GetAuthReturn =
+  | AuthObjectWithoutResources<SignedInAuthObject>
+  | AuthObjectWithoutResources<SignedOutAuthObject>;
+
+export const createGetAuth = ({ noAuthStatusMessage }: { noAuthStatusMessage: string }) => {
+  return (
+    req: Request,
+    locals: APIContext['locals'],
+    opts?: { secretKey?: string },
+  ): AuthObjectWithoutResources<SignedInAuthObject> | AuthObjectWithoutResources<SignedOutAuthObject> => {
+    // When the auth status is set, we trust that the middleware has already run
+    // Then, we don't have to re-verify the JWT here,
+    // we can just strip out the claims manually.
+    const authToken = locals.authToken || getAuthKeyFromRequest(req, 'AuthToken');
```

Comment:
how does Astro middleware work? Does it run in the same process as the app server itself, or is it like Next where it runs somewhere else? Do we need to worry about the validity of the token here or are we safe to assume it is the token we previously verified in the middleware?

Reply:
It runs on the same machine, where SSR takes place. Astro supports deployment to vercel where you can choose to use the edge middleware, but we are kind of blocking that because our middleware is using an AsyncLocalStorage